File: components/platform/webview/src/js/bridge.js
// The WaterUI JavaScript bridge.
//
// One copy, shared by every backend. Each backend supplies only a transport: a
// function that hands `__wateruiSend` a request envelope as a string, and a call
// back into `__wateruiResolve` with the reply.
//
// Page-facing API:
//
//   const reply = await waterui.invoke("greet", { name: "Lexo" });  // → { text: "Hi Lexo" }
//   const bytes = await waterui.invoke("read", "/a.txt");           // → Uint8Array
//
// One call for both: what comes back is decided by what the handler returns,
// which the page cannot influence. There was a second entry point that asked
// for bytes, but a handler returning a value could not honour it and one
// returning bytes needed no asking.
//
// `waterui.invoke` is namespaced deliberately: an earlier design installed each
// handler as `window[name]`, which silently overwrote page globals whenever a
// handler was called `open`, `name`, `print`, and so on.
(function () {
  if (globalThis.waterui !== undefined) {
    return;
  }

  var pending = new Map();

  // Ids are unique to this document, not just within it.
  //
  // A handler may still be awaiting I/O when the page navigates. The reply is
  // then evaluated in the *new* document, whose own `pending` map started over
  // from the same low numbers — so a stale reply could settle an unrelated call
  // with another call's value. Starting each document at a random point makes
  // an id from a previous document essentially certain not to be pending here,
  // and `__wateruiResolve` already ignores ids it does not know.
  //
  // 26 bits of document entropy and 26 bits of counter stay well inside the
  // range a double holds exactly, so the id crosses as an ordinary number.
  var nextId = Math.floor(Math.random() * 0x4000000) * 0x4000000;

  // Integers Rust can hold and a JS number cannot. JSON has no spelling for
  // them, so they cross as {"__wateruiBigInt": "123..."} and become BigInt
  // here; anything that fits in a double is an ordinary number and stays one,
  // so page arithmetic on normal values is unaffected. Shared with state.js
  // through the internal global below.
  var BIG_INT_TAG = "__wateruiBigInt";

  function reviveBigInts(value) {
    if (value === null || typeof value !== "object") {
      return value;
    }
    if (Array.isArray(value)) {
      for (var i = 0; i < value.length; i += 1) {
        value[i] = reviveBigInts(value[i]);
      }
      return value;
    }
    var keys = Object.keys(value);
    if (keys.length === 1 && keys[0] === BIG_INT_TAG && typeof value[BIG_INT_TAG] === "string") {
      return BigInt(value[BIG_INT_TAG]);
    }
    for (var k = 0; k < keys.length; k += 1) {
      value[keys[k]] = reviveBigInts(value[keys[k]]);
    }
    return value;
  }

  // JSON.stringify throws on a BigInt rather than guessing, so every value
  // leaving the page goes through this.
  function replaceBigInts(key, value) {
    if (typeof value === "bigint") {
      var tagged = {};
      tagged[BIG_INT_TAG] = value.toString();
      return tagged;
    }
    return value;
  }

  globalThis.__wateruiBigInts = {
    revive: reviveBigInts,
    replacer: replaceBigInts,
  };

  function toBase64(bytes) {
    var binary = "";
    var view = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
    for (var i = 0; i < view.length; i += 1) {
      binary += String.fromCharCode(view[i]);
    }
    return btoa(binary);
  }

  function fromBase64(text) {
    var binary = atob(text);
    var bytes = new Uint8Array(binary.length);
    for (var i = 0; i < binary.length; i += 1) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }

  // Called by the native side exactly once per request. An unknown id is ignored
  // rather than thrown: page script can reach this function, and a stray call must
  // not surface as an unhandled exception.
  globalThis.__wateruiResolve = function (id, ok, payload) {
    var entry = pending.get(id);
    if (entry === undefined) {
      return;
    }
    pending.delete(id);
    if (!ok) {
      var error = new Error(String((payload && payload.message) || "WaterUI handler failed"));
      if (payload && payload.data !== undefined) {
        error.data = payload.data;
      }
      entry.reject(error);
      return;
    }
    // Which branch runs is the handler's choice, not the caller's: a handler
    // that returns bytes sends `b64`, one that returns a value sends `json`.
    // Base64 is how bytes travel, never the answer itself, so it is always
    // decoded.
    if (payload && typeof payload.b64 === "string") {
      entry.resolve(fromBase64(payload.b64));
      return;
    }
    entry.resolve(payload ? reviveBigInts(payload.json) : undefined);
  };

  function send(name, payload) {
    var id = (nextId += 1);
    return new Promise(function (resolve, reject) {
      pending.set(id, { resolve: resolve, reject: reject });
      var envelope = { id: id, name: String(name) };
      if (payload instanceof Uint8Array || payload instanceof ArrayBuffer) {
        envelope.b64 = toBase64(payload);
      } else {
        // JSON is the common path and stays out of base64: it is what the engine's
        // own JSON.stringify produces and what the native side parses directly.
        envelope.json = payload === undefined ? null : payload;
      }
      globalThis.__wateruiSend(JSON.stringify(envelope, replaceBigInts));
    });
  }

  globalThis.waterui = {};

  // Not writable and not configurable, so page script cannot swap the transport
  // out from under the handlers — but the object itself stays extensible.
  // Freezing it whole is what this used to do, and `state.js` runs after this
  // file and adds `state` and `watch` to the same object: `defineProperty` on a
  // frozen object throws, so neither has ever existed in a page and the entire
  // mirrored-state feature was unreachable from JavaScript.
  Object.defineProperty(globalThis.waterui, "invoke", {
    value: function (name, payload) {
      return send(name, payload);
    },
    enumerable: true,
  });
})();

File: components/platform/webview/src/bridge.rs
//! The JavaScript bridge shared by every `WebView` backend.
//!
//! Four backends each carried a near-identical copy of the bridge script with a
//! subtly different wire format, and two more (CEF, WPE) exposed a different
//! page-facing API entirely. A snippet written against one backend could not run
//! on another, which defeats the purpose of having one `WebView` component.
//!
//! This module owns the single script and the single envelope. A backend supplies
//! only a transport: deliver [`SCRIPT`] at document start, hand
//! [`Request::parse`] whatever `__wateruiSend` produced, and evaluate
//! [`Reply::resolve_script`] to complete the call.

use base64::Engine as _;
use base64::engine::general_purpose::STANDARD;
use serde::{Deserialize, Serialize};

/// The bridge script, injected at document start by every backend.
pub const SCRIPT: &str = include_str!("js/bridge.js");

/// The global the bridge calls to hand a request to the native side.
///
/// A backend installs a function under this name — a `WebKit` message handler, an
/// Android `@JavascriptInterface` method, a CDP binding — taking one string.
pub const SEND_FUNCTION: &str = "__wateruiSend";

/// One `waterui.invoke(...)` call, as parsed from the page.
///
/// Every field is validated: this arrives from page script, which may be
/// untrusted, so a malformed request is rejected rather than fatal.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Request {
    /// Correlates the reply with the pending promise in the page.
    pub id: u64,
    /// The handler name the page asked for.
    pub name: String,
    /// The payload, already decoded to bytes.
    pub payload: Vec<u8>,
    /// Whether the payload arrived as binary rather than JSON.
    pub is_binary: bool,
}

#[derive(Debug, Deserialize)]
struct RawRequest {
    id: u64,
    name: String,
    #[serde(default)]
    json: Option<serde_json::Value>,
    #[serde(default)]
    b64: Option<String>,
}

/// Why a request from the page could not be understood.
#[derive(Debug, thiserror::Error)]
pub enum RequestError {
    /// The envelope was not the JSON object the bridge produces.
    #[error("malformed WaterUI bridge envelope: {0}")]
    Envelope(#[from] serde_json::Error),
    /// The binary payload was not valid base64.
    #[error("malformed WaterUI bridge payload: {0}")]
    Payload(#[from] base64::DecodeError),
}

impl Request {
    /// Parses the envelope a backend received from [`SEND_FUNCTION`].
    ///
    /// # Errors
    ///
    /// Returns an error when the envelope is not well-formed. Reject the call
    /// with [`Reply::failure`] rather than treating it as fatal: page script can
    /// reach the transport directly.
    pub fn parse(envelope: &str) -> Result<Self, RequestError> {
        let raw: RawRequest = serde_json::from_str(envelope)?;
        let (payload, is_binary) = match (raw.json, raw.b64) {
            (_, Some(encoded)) => (STANDARD.decode(encoded)?, true),
            (Some(value), None) => (serde_json::to_vec(&value)?, false),
            (None, None) => (Vec::new(), false),
        };
        Ok(Self {
            id: raw.id,
            name: raw.name,
            payload,
            is_binary,
        })
    }
}

/// The result of a handler, ready to be delivered back to the page.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Reply {
    /// Bytes the handler produced, delivered as base64.
    Bytes(Vec<u8>),
    /// The call failed; the page's promise rejects with this message.
    Failure(String),
}

#[derive(Serialize)]
#[serde(untagged)]
enum ReplyPayload<'a> {
    Bytes { b64: String },
    Failure { message: &'a str },
}

impl Reply {
    /// Builds a failure reply.
    pub fn failure(message: &(impl ToString + ?Sized)) -> Self {
        Self::Failure(message.to_string())
    }

    /// Renders the JavaScript that completes the pending call.
    ///
    /// Both arguments are serialized rather than interpolated, so a handler's
    /// bytes or an error message can never be read as source.
    ///
    /// # Panics
    ///
    /// Panics if the reply cannot be serialized, which would mean a bug in this
    /// module rather than anything the page did.
    #[must_use]
    pub fn resolve_script(&self, id: u64) -> String {
        let (ok, payload) = match self {
            Self::Bytes(bytes) => (
                true,
                ReplyPayload::Bytes {
                    b64: STANDARD.encode(bytes),
                },
            ),
            Self::Failure(message) => (false, ReplyPayload::Failure { message }),
        };
        let payload =
            serde_json::to_string(&payload).expect("WaterUI bridge reply must serialize");
        format!("globalThis.__wateruiResolve({id},{ok},{payload});")
    }
}

#[cfg(test)]
mod tests {
    use super::{Reply, Request};

    #[test]
    fn a_json_payload_arrives_without_base64() {
        let request = Request::parse(r#"{"id":7,"name":"greet","json":{"name":"Lexo"}}"#)
            .expect("well-formed envelope");

        assert_eq!(request.id, 7);
        assert_eq!(request.name, "greet");
        assert!(!request.is_binary);
        assert_eq!(request.payload, br#"{"name":"Lexo"}"#);
    }

    #[test]
    fn a_binary_payload_round_trips_through_base64() {
        let request =
            Request::parse(r#"{"id":1,"name":"upload","b64":"AAEC"}"#).expect("well-formed");

        assert!(request.is_binary);
        assert_eq!(request.payload, vec![0, 1, 2]);
    }

    #[test]
    fn a_call_with_no_payload_is_accepted() {
        let request = Request::parse(r#"{"id":2,"name":"ping"}"#).expect("well-formed");
        assert!(request.payload.is_empty());
    }

    /// Page script can reach the transport directly, so these must be errors
    /// rather than panics.
    #[test]
    fn malformed_envelopes_are_rejected_not_fatal() {
        assert!(Request::parse("garbage").is_err());
        assert!(Request::parse(r#"{"id":"not a number","name":"x"}"#).is_err());
        assert!(Request::parse(r#"{"name":"missing id"}"#).is_err());
        assert!(Request::parse(r#"{"id":1,"name":"x","b64":"not base64!"}"#).is_err());
    }

    #[test]
    fn replies_serialize_their_payload_rather_than_interpolating_it() {
        let script = Reply::Bytes(vec![1, 2, 3]).resolve_script(4);
        assert_eq!(
            script,
            r#"globalThis.__wateruiResolve(4,true,{"b64":"AQID"});"#
        );

        // A message containing quotes and a closing brace must not escape the call.
        let script = Reply::failure(r#"he said "}"; alert(1)"#).resolve_script(5);
        assert_eq!(
            script,
            r#"globalThis.__wateruiResolve(5,false,{"message":"he said \"}\"; alert(1)"});"#
        );
    }
}
